apex/contrib/csrc/cudnn_gbn/cudnn_gbn.cpp

```diff
@@ -38,28 +44,43 @@ at::Tensor gbn_forward(const at::Tensor& x,
     void_peer_buffers.push_back((void*)addr);
   }
 
+  // we need the peer size for the buffer reset
+  size_t peer_size = 1;
+  for (size_t i = 0; i < 4; ++i){
+    peer_size *= peerDims[i];
+  }
+
+  // sanity check
   assert(bn_group == void_peer_buffers.size());
-  run_batch_norm_forward(
-    perChannelDims,
-    epsilonDims,
-    tensorDims,
-    peerDims,
-    x.data_ptr(),
-    y.data_ptr(),
-    scale.data_ptr(),
-    bias.data_ptr(),
-    running_mean.data_ptr(),
-    running_var.data_ptr(),
-    running_mean.data_ptr(),
-    running_var.data_ptr(),
-    minibatch_mean.data_ptr(),
-    minibatch_inv_var.data_ptr(),
-    void_peer_buffers,
-    epsilon,
-    momentum,
-    rank_id
-  );
 
+  // check if plan already exists
+  std::vector<int64_t> fv = {(int64_t)BN_FWD, N, C, H, W, bn_group, (int64_t)CUDNN_DATA_HALF};
+  if ( gbn_plan_cache.find(fv) == gbn_plan_cache.end() ) {
+    auto plan = run_batch_norm_forward(tensorDims, perChannelDims, epsilonDims, peerDims, CUDNN_DATA_HALF);
+    gbn_plan_cache.insert(std::make_pair(fv, plan));
+  }
+
```

Comment:
It looks like some of the code makes assumptions about the input tensor(s)' memory layout. If so, there should be checks like `is_contiguous(at::MemoryFormat::ChannelsLast)`.

Reply:
This is done on the python frontend. That check it [here](https://github.com/azrael417/apex/blob/96b961f38128cce0452d53c8b501e28e220b28e2/apex/contrib/cudnn_gbn/batch_norm.py#L122)